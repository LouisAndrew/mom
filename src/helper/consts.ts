export const clickToChatLink = 'https://wa.me/6287782832988';
