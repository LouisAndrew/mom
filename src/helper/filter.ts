import { intersection, flatten } from 'lodash';

import { Property } from 'interfaces/Property';
import { MinMaxObj } from 'templates/products/products';

type FilterObj = {
    addressFilter: string;
    locationFilters: string[];
    saleTypeFilters: string[];
    propTypeFilters: string[];
    areaFilter: MinMaxObj;
    priceFilter: MinMaxObj;
    properties: Property[];
};

const filterByAddress = (addr: string, properties: Property[]): Property[] =>
    properties.filter(
        property =>
            property.address.toLowerCase().indexOf(addr.toLowerCase()) !== -1
    );

const filterByLocation = (
    locations: string[],
    properties: Property[]
): Property[] =>
    flatten(
        locations.map(location =>
            properties.filter(property => property.location === location)
        )
    );

const filterBySaleType = (
    saleTypes: string[],
    properties: Property[]
): Property[] =>
    flatten(
        saleTypes.map(saleType =>
            properties.filter(property => property.saleType === saleType)
        )
    );

const filterByPropType = (
    propTypes: string[],
    properties: Property[]
): Property[] =>
    flatten(
        propTypes.map(propType =>
            properties.filter(property => property.propertyType === propType)
        )
    );

const filterByArea = (area: MinMaxObj, properties: Property[]): Property[] =>
    properties.filter(
        property =>
            property.area &&
            property.area >= area.min &&
            property.area <= area.max
    );

const filterByPrice = (price: MinMaxObj, properties: Property[]): Property[] =>
    properties.filter(
        property =>
            property.price &&
            property.price >= price.min &&
            property.price <= price.max
    );

/**
 * filters properties given by all filter properties.
 */
export const filter = ({
    addressFilter,
    locationFilters,
    saleTypeFilters,
    propTypeFilters,
    areaFilter,
    priceFilter,
    properties,
}: FilterObj): Promise<Property[]> => {
    const filteredByAddress = filterByAddress(addressFilter, properties);
    // just use the filter if any tags is provided.
    const filteredByLocation =
        locationFilters.length > 0
            ? filterByLocation(locationFilters, properties)
            : properties;
    const filteredBySaleType =
        saleTypeFilters.length > 0
            ? filterBySaleType(saleTypeFilters, properties)
            : properties;
    const filteredByPropType =
        propTypeFilters.length > 0
            ? filterByPropType(propTypeFilters, properties)
            : properties;
    //
    const filteredByArea = filterByArea(areaFilter, properties);
    const filteredByPrice = filterByPrice(priceFilter, properties);

    // returns a promise to provide the ability to use any loader.
    return new Promise<Property[]>((res, rej) => {
        res(
            intersection(
                properties,
                filteredByAddress,
                filteredBySaleType,
                filteredByLocation,
                filteredByPropType,
                filteredByArea,
                filteredByPrice
            )
        );
    });
};
