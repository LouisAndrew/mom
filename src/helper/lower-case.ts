export const createSlug = (name: string): string =>
    name
        .toLowerCase()
        .split(' ')
        .join('-');
