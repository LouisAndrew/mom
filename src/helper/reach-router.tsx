import React from 'react';
import { LocationProvider } from '@reach/router';

const withReachRouter = (el: any) => <LocationProvider>{el}</LocationProvider>;

export { withReachRouter };
