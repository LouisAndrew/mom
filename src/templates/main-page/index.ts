import { MainPage } from './main-page';

export default MainPage;

// TODO: create page component here!
