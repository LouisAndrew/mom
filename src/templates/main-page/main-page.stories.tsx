import React from 'react';
import { storiesOf } from '@storybook/react';
import MainPage from '.';
// import more addons

const story = storiesOf('Main Page', module);

story.add('Component', () => <MainPage />);
