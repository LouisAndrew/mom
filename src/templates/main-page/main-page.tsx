import React from 'react';

import Hero from './hero';
import About from './about';
import HotList from './hot-list';
import { Property } from 'interfaces/Property';
import { FluidObject } from 'gatsby-image';

type Props = {
    heroImgFluid?: FluidObject;
    aboutImgFluid?: FluidObject;
    hotProperties?: Property[];
};

// TODO: gatsby context handling!
const MainPage: React.FC<Props> = ({
    heroImgFluid,
    aboutImgFluid,
    hotProperties,
}) => {
    const mockProperties: Property[] = [
        {
            name: 'Rumah 1',
            address: 'jalan Mangga 24 blok G nomor 167',
            imgs: [
                {
                    image:
                        'https://res.cloudinary.com/dsvdffre0/image/upload/v1595340453/ovmhcixqqpfjla4mxovf.jpg',
                    imgAlt: 'Handsome boy',
                },
            ],
            propertyType: 'house',
            saleType: 'sell',
        },
        {
            name: 'Rumah 2',
            address: 'jalan Mangga 25 blok F nomor 167',
            imgs: [
                {
                    image:
                        'https://res.cloudinary.com/dsvdffre0/image/upload/v1596838654/felix-lam-J7fxkhtOqt0-unsplash_wcncna.jpg',
                    imgAlt: 'Lakers logo',
                },
                {
                    image:
                        'https://res.cloudinary.com/dsvdffre0/image/upload/v1595015542/sample.jpg',
                    imgAlt: 'Flower',
                },
            ],
            propertyType: 'apartment',
            saleType: 'rent',
        },
        {
            name: 'Apartment 2',
            address: 'jalan Mangga 25 blok F nomor 167',
            imgs: undefined,
            propertyType: 'apartment',
            saleType: 'rent',
        },
    ];

    const src =
        'https://res.cloudinary.com/dsvdffre0/image/upload/v1596841161/stefan-stefancik-QXevDflbl8A-unsplash_ei8bvp.jpg';
    const imgAlt = 'smiling-woman';

    return (
        <>
            <Hero heroImgFluid={heroImgFluid} />
            <About imgSrc={src} imgFluidSrc={aboutImgFluid} imgAlt={imgAlt} />
            <HotList
                hotProperties={hotProperties ? hotProperties : mockProperties}
            />
        </>
    );
};

export { MainPage };
