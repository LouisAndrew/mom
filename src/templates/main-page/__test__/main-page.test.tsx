import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';

import { MainPage } from '../main-page';
import { withReachRouter } from 'helper/reach-router';

import * as Gatsby from 'gatsby';

describe('Main page.', () => {
    const useStaticQuery = jest.spyOn(Gatsby, 'useStaticQuery');

    beforeEach(() => {
        useStaticQuery.mockImplementationOnce(() => {
            return {
                allSanityLocation: {
                    edges: [
                        {
                            node: {
                                locationName: 'Mock name',
                            },
                        },
                    ],
                },
            };
        });
    });

    afterEach(cleanup);
    const el: React.ReactElement = withReachRouter(
        <MainPage />
    ) as React.ReactElement;

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(el, div);
    });

    it('matches snapshot', () => {
        const tree = renderer.create(el).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
