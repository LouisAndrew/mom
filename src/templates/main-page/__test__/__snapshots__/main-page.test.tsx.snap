// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main page. matches snapshot 1`] = `null`;
